app/src/main/java/org/mozilla/focus/web/GeckoWebViewProvider.kt

```diff
@@ -636,8 +644,8 @@ class GeckoWebViewProvider : IWebViewProvider {
         }
 
         override fun saveWebViewState(session: Session) {
+            session.savedGeckoSession = geckoSession
             val sessionBundle = Bundle()
-            sessionBundle.putParcelable(GECKO_SESSION, geckoSession)
```

Comment:
Might be missing something, but wouldn't we have to listen to `onSessionStateChange` and still add the `SessionState.toString` to the Bundle and restore it later? Otherwise we may lose the session on pause/resume...

Reply:
From what I read [here](https://searchfox.org/mozilla-central/source/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/GeckoSession.java#2458-2479), the `SessionState` only stores the history, form and scroll data for the session. Focus wants to hold on to state temporarily when clicking the 'Open in Focus' in custom tabs and when putting the app into the background. Adding the whole gecko session to the `SessionExtensions` seemed to be a better solution to hold the session without needing to add/restore the `SessionState`.

The existing way for this is with the `SessionExtensions` where the session was stored in a bundle. I added the `savedGeckoSession` item in the `SessionExtensions` without the serializing.

Sebastian raised a concern about that above about losing this session when doing this that I have to address which may change this.